### PR TITLE
[inductor] Don't inherit `__future__` flags from the calling scope when `compile` -ing generated modules

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126098
* #126195
* __->__ #126454

This file includes `from __futures__ import annotations` which interacts with `compile` by causing type annotations to be populated as strings. Triton does not parse the string annotation correctly. Avoid this behavior by passing `dont_inherit=True` to `compile`.  

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/runtime/compile_tasks.py [hunk 1]
```diff
@@ -1,11 +1,10 @@
-from __future__ import annotations
```
Comment (peterbell10) [inline torch/_inductor/runtime/compile_tasks.py:1]: We should be able to keep this now, right?

Comment (amjames) [reply torch/_inductor/runtime/compile_tasks.py:1]: Yes, but now that we are not using `compile` it in fact does not do anything. I see no reason to keep it.

Comment (peterbell10) [reply torch/_inductor/runtime/compile_tasks.py:1]: It wasn't added for the purpose of compile, so I don't see any reason to remove it?

Comment (amjames) [reply torch/_inductor/runtime/compile_tasks.py:1]: Fair enough. I will add it back in.

**Review by peterbell10 — APPROVED** [review]
Looks good, modulo the lint failures

**Comment by DanilBaibak** [comment]
Hi @amjames! Sorry, I need to revert your PR because it breaks the internal build. Here are the details:
```
4 TESTS FAILED
  ✗ caffe2/test/inductor:multi_kernel - test_reduction_scratch_buffer_cpp_wrapper (caffe2.test.inductor.test_multi_kernel.MultiKernelTest)
  ✗ caffe2/test/inductor:multi_kernel - test_reduction_scratch_buffer_cpp_wrapper_non_persistent_reduction (caffe2.test.inductor.test_multi_kernel.MultiKernelTest)
  ✗ caffe2/test/inductor:multi_kernel - test_reduction_scratch_buffer_cpp_wrapper_persistent_reduction (caffe2.test.inductor.test_multi_kernel.MultiKernelTest)
  ✗ caffe2/test/inductor:multi_kernel - test_softmax_cpp_wrapper (caffe2.test.inductor.test_multi_kernel.MultiKernelTest)
  ```
  
  Traceback:
  ```
  [2024-05-19T08:42:33.790-07:00]   File "/dev/shm/uid-30083/e06234a2-seed-nspid4026534425_cgpid3536527-ns-4026534755/torch/_inductor/graph.py", line 1738, in compile_to_fn
[2024-05-19T08:42:33.790-07:00]     return self.compile_to_module().call
[2024-05-19T08:42:33.790-07:00]   File "/dev/shm/uid-30083/e06234a2-seed-nspid4026534425_cgpid3536527-ns-4026534755/torch/_inductor/utils.py", line 1018, in patched_compile_to_module
[2024-05-19T08:42:33.790-07:00]     mod = compile_to_module(self)
[2024-05-19T08:42:33.790-07:00]   File "/dev/shm/uid-30083/e06234a2-seed-nspid4026534425_cgpid3536527-ns-4026534755/torch/_dynamo/utils.py", line 210, in time_wrapper
[2024-05-19T08:42:33.790-07:00]     r = func(*args, **kwargs)
[2024-05-19T08:42:33.790-07:00]   File "/dev/shm/uid-30083/e06234a2-seed-nspid4026534425_cgpid3536527-ns-4026534755/torch/_inductor/graph.py", line 1682, in compile_to_module
[2024-05-19T08:42:33.790-07:00]     mod = PyCodeCache.load_by_key_path(
[2024-05-19T08:42:33.790-07:00]   File "/dev/shm/uid-30083/e06234a2-seed-nspid4026534425_cgpid3536527-ns-4026534755/torch/_inductor/codecache.py", line 2584, in load_by_key_path
[2024-05-19T08:42:33.790-07:00]     mod = _reload_python_module(key, path)
[2024-05-19T08:42:33.790-07:00]   File "/dev/shm/uid-30083/e06234a2-seed-nspid4026534425_cgpid3536527-ns-4026534755/torch/_inductor/runtime/compile_tasks.py", line 42, in _reload_python_module
[2024-05-19T08:42:33.790-07:00]     raise RuntimeError(
[2024-05-19T08:42:33.790-07:00] torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
[2024-05-19T08:42:33.790-07:00] RuntimeError: Failed to import /re_tmp/tmppyq1p33x/ca/ccaifp7q25wp7vqbispli5ky6lo5b2qq3zoa75eped3jjeyvrncj.py
[2024-05-19T08:42:33.790-07:00] SkipFrame: COMPILE FAILED WITH EXIT CODE 1
[2024-05-19T08:42:33.790-07:00] stdout: 
[2024-05-19T08:42:33.790-07:00] stderr: 
[2024-05-19T08:42:33.790-07:00]
[2024-05-19T08:42:33.790-07:00] Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
[2024-05-19T08:42:33.790-07:00]
[2024-05-19T08:42:33.790-07:00]
[2024-05-19T08:42:33.790-07:00] You can suppress this exception and fall back to eager by setting:
[2024-05-19T08:42:33.790-07:00]     import torch._dynamo
[2024-05-19T08:42:33.790-07:00]     torch._dynamo.config.suppress_errors = True
  ```

**Comment by amjames** [comment]
I reverted back to using `compile` + `exec` here, the original issue can be resolved using the `dont_inherit` kwarg to compile so that is not so bad.


@DanilBaibak do you want to see if those tests fail on the internal build, or should I just try to re-land this after CI is done?

**Comment by DanilBaibak** [comment]
@amjames, I think you can just re-land after CI is done.

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**